assert: fix _deepEqual and improve assert.md

joyeecheung opened the pull request:
This is the more semver-patch part of https://github.com/nodejs/node/pull/10282.

### Fix _deepEqual

Refactors _deepEqual and fixes a few code paths that lead to
behaviors contradicting what the doc says. Before this commit
certain types of objects (Buffers, Dates, etc.) are not checked
properly, and can get away with different prototypes AND different
enumerable owned properties because _deepEqual would jump to
premature conclusion for them.

Since we no longer follow CommonJS unit testing spec,
the checks for primitives and object prototypes are moved
forward for faster failure.

Improve regexp and float* array checks:

* Don't compare lastIndex of regexps, because they are not
  enumerable, so according to the docs they should not be compared
* Compare flags of regexps instead of separate properties
* Use built-in tags to test for float* arrays instead of using
  instanceof

Refs: https://github.com/nodejs/node/pull/10282#issuecomment-274267895
Refs: https://github.com/nodejs/node/issues/10258#issuecomment-266963234

### Improve documentation

Use the term "Abstract Equality Comparison" and "Strict Equality
Comparison" from ECMAScript specification to refer to the operations
done by `==` and `===`, instead of the informal
"equal comparison operator" and "strict equality operator".

Clarify that deep strict comparisons checks `[[Prototype]]`
property, instead of the vague "object prototypes".

Suggest using `Object.is()` to avoid the caveats of +0, -0 and NaN.

Add a MDN link explaining what enumerable "own" properties are.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
test, assert, doc

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/562/

joyeecheung commented:
Benchmarks results (benchmarks taken from https://github.com/nodejs/node/pull/11092). Note that this implementation favors fast failures, though failures are not benchmarked(when an assertion fails the process should probably be terminated anyway).

<details>
<summary>see results</summary>
<pre>
                                                                                                  improvement confidence      p.value
 assert/deepequal-buffer.js method="nonstrict" len=100 n=1000                                        -92.69 %        *** 1.398352e-10
 assert/deepequal-buffer.js method="strict" len=100 n=1000                                           -87.27 %        *** 8.988908e-07
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="array"          0.52 %            5.671380e-01
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="boolean"        1.33 %            3.162478e-01
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="new-array"      1.39 %          * 2.256608e-02
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="null"           1.12 %            1.546353e-01
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="number"        -0.52 %            7.352546e-01
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="object"        -0.53 %            5.363293e-01
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="string"         0.48 %            5.514581e-01
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="undefined"      0.70 %            2.601707e-01
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="array"             0.59 %            4.341103e-01
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="boolean"           1.73 %         ** 9.211000e-03
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="new-array"         1.30 %            1.185455e-01
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="null"              0.68 %            3.269869e-01
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="number"           -2.51 %            5.382094e-02
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="object"            2.23 %            8.626440e-02
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="string"            0.20 %            7.698036e-01
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="undefined"         0.86 %            3.789055e-01
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="array"                 6.32 %        *** 4.516185e-04
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="boolean"               9.25 %        *** 8.115494e-05
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="new-array"             9.36 %        *** 7.678773e-05
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="null"                  9.41 %        *** 2.294144e-04
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="number"                5.07 %          * 3.376728e-02
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="object"                4.23 %            1.091126e-01
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="string"                6.42 %        *** 5.440464e-04
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="undefined"             8.15 %         ** 4.303777e-03
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="array"                    6.60 %          * 1.017124e-02
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="boolean"                  8.82 %        *** 6.257818e-05
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="new-array"               11.13 %        *** 6.845735e-04
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="null"                     8.88 %        *** 1.083174e-04
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="number"                   4.29 %            2.310841e-01
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="object"                   7.70 %        *** 2.296652e-04
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="string"                   9.81 %        *** 1.067824e-06
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="undefined"               13.05 %        *** 1.293929e-05
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Float32Array"                0.45 %            6.665986e-01
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Float64Array"                1.62 %            1.182916e-01
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Int16Array"                -98.85 %        *** 6.838162e-11
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Int32Array"                -98.02 %        *** 1.541802e-11
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Int8Array"                 -99.33 %        *** 1.169792e-12
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Uint16Array"               -98.88 %        *** 1.836848e-11
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Uint32Array"               -97.99 %        *** 7.248263e-13
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Uint8Array"                -99.32 %        *** 3.145163e-11
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Uint8ClampedArray"         -99.30 %        *** 5.984442e-12
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Float32Array"                   0.78 %            4.221460e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Float64Array"                  -0.08 %            9.232714e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Int16Array"                    -6.42 %            7.496064e-02
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Int32Array"                    -0.20 %            9.728131e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Int8Array"                     -5.38 %            5.518969e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Uint16Array"                   -6.58 %            2.480984e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Uint32Array"                   -4.62 %            1.772045e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Uint8Array"                   -10.72 %        *** 1.283864e-04
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Uint8ClampedArray"             -0.45 %            8.987984e-01
</pre>
</details>

joyeecheung commented:
Hmm, canary died but I can't be sure if the failures are related(most are in network-related modules, and node-sass has an error about platform support). @MylesBorins Am I using it properly?

jasnell commented on doc/api/assert.md:
```diff
@@ -40,13 +40,18 @@ added: v0.1.21
+
+Only [enumerable "own" properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties)
+are considered. The `deepEqual()` implementation does not test the
+[`[[Prototype]]`](https://tc39.github.io/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots)
```
Please move the link ref to the end of the file and use an inline ref like:

```
[`[[Prototype]]`][]
```

joyeecheung replied:
I used this bug seems GitHub has a little bit trouble with that(I am guessing it's the `[[` and `]]` somehow confuses its parser?), not sure if the tooling can deal with it?

jasnell commented on doc/api/assert.md:
```diff
@@ -40,13 +40,18 @@ added: v0.1.21
+Only [enumerable "own" properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties)
+are considered. The `deepEqual()` implementation does not test the
+[`[[Prototype]]`](https://tc39.github.io/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots)
+of objects, attached symbols, or non-enumerable
```
This line is rather short, can you update this so that the line length is closer to 80

jasnell commented on doc/api/assert.md:
```diff
@@ -469,6 +478,33 @@ assert.throws(myFunction, 'missing foo', 'did not throw with expected message');
+
+For the following cases, consider using ES2015 [`Object.is()`][],
+which uses the
+[SameValueZero](https://tc39.github.io/ecma262/#sec-samevaluezero) comparison.
```
Please move this to a reference at the bottom of the file also. :-)

jasnell commented on test/parallel/test-assert-deep.js:
```diff
@@ -0,0 +1,109 @@
+    result += str.replace(/[\\^$.*+?()[\]{}|=!<>:-]/g, '\\$&');
+    result += literals[i + 1];
+  }
+  return new RegExp('^AssertionError: ' + result + '$');
```
```js
return new RegExp(`^AssertionError: ${result}$`);
```

joyeecheung commented:
@jasnell Thanks, PTAL. I am not quite sure about what to do the `[[Prototype]]` part, because neither the markdown rendering of GitHub nor our tooling can deal with them:

<img width="945" alt="screen shot 2017-02-04 at 2 29 21 am" src="https://cloud.githubusercontent.com/assets/4299420/22603259/a00c7b88-ea81-11e6-8573-e9b17aaca7dc.png">

jasnell commented:
Escape the `[` with a `\` like so:
```
\`[\[Prototype\]\`][]

\`[\[Prototype\]\`]: https://example.org/link/to/whatever
```

For instance: [`\[\[Prototype\]\]`][]

[`\[\[Prototype\]\]`]: https://example.org/link/to/whatever

joyeecheung commented:
Hmm...putting it inside `` ` `` still doesn't work:

```
[`\[\[Prototype\]\]`]: https://tc39.github.io/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots
```

<img width="938" alt="screen shot 2017-02-04 at 4 00 59 pm" src="https://cloud.githubusercontent.com/assets/4299420/22616911/f55f847e-eaf2-11e6-8e00-40673cf6e1cf.png">

joyeecheung commented:
@addaleax @Trott Can you take a look at this if you have time since you've LTGM'ed https://github.com/nodejs/node/pull/10282? Thanks?

ljharb commented on lib/assert.js:
```diff
@@ -140,117 +139,176 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
+}
+
+function isArguments(tag) {
+  return tag == '[object Arguments]';
```
`===`

ljharb commented on lib/assert.js:
```diff
@@ -140,117 +139,176 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
+  // for a list of tags pre-defined in the spec.
+  // There are some unspecified tags in the wild too (e.g. typed array tags).
+  // Since tags can be altered, they only serve fast failures
+  const actualTag = Object.prototype.toString.call(actual);
```
Same comment as I made in https://github.com/nodejs/node/pull/10282 - `Object.prototype.toString` should be cached at module level.

joyeecheung commented:
Rebased and, ah...right, I was being silly, the `[[Prototype]]` link can be easily fixed by using named reference links :P  @jasnell PTAL.

joyeecheung commented:
Pinging @jasnell @Trott @addaleax for more reviews, thanks!

jasnell commented on doc/api/assert.md:
```diff
@@ -33,13 +33,17 @@ added: v0.1.21
+Primitive values are compared with the [Abstract Equality Comparison][]
+( `==` ).
+
+Only [enumerable "own" properties](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties)
```
Please move the link to the bottom and use an inline reference like `[enumerable "own" properties][]` to avoid having the over-long line

jasnell commented on doc/api/assert.md:
```diff
@@ -33,13 +33,17 @@ added: v0.1.21
+[`[[Prototype]]`][prototype-spec] of objects, attached symbols, or
+non-enumerable properties (for those checks, consider using
+[`assert.deepStrictEqual()`][] instead). This can lead to some
+potentially surprising results. For example, the following example does no
```
s/does no/does not (end of the line)

jasnell commented on doc/api/assert.md:
```diff
@@ -462,10 +470,43 @@ assert.throws(myFunction, 'missing foo', 'did not throw with expected message');
+
+For the following cases, consider using ES2015 [`Object.is()`][],
+which uses the
+[SameValueZero][] comparison.
```
Move this up to the previous line :-)

jasnell commented on doc/api/assert.md:
````diff
@@ -462,10 +470,43 @@ assert.throws(myFunction, 'missing foo', 'did not throw with expected message');
+```
+
+For more information, see
+[MDN's guide on equality comparisons and sameness](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness).
````
Even tho this is at the end of the file, please move the link to the index at the end and use an inline reference for this as well, just for consistency :-)

joyeecheung commented:
@jasnell Updated, also rebased against master to make sure `make doc` looks right....PTAL, thanks.

Trott commented on doc/api/assert.md:
```diff
@@ -33,13 +33,16 @@ added: v0.1.21
-because the properties on the [`Error`][] object are non-enumerable:
+Only [enumerable "own" properties][] are considered. The `deepEqual()`
+implementation does not test the [`[[Prototype]]`][prototype-spec] of
+objects, attached symbols, or non-enumerable properties (for those checks, 
```
Punctuation nit that can be ignored: 

> objects, attached symbols, or non-enumerable properties. (For those checks, 
> consider using [`assert.deepStrictEqual()`][] instead.)

joyeecheung commented:
Punctuation fixed, thanks! @Trott 

CI: https://ci.nodejs.org/job/node-test-pull-request/6473/

joyeecheung commented:
Recent CITGM jobs are all red, launched https://ci.nodejs.org/job/citgm-smoker/582/ against master and https://ci.nodejs.org/job/citgm-smoker/583/ against this PR for comparison.

jasnell commented on doc/api/assert.md:
```diff
@@ -33,13 +33,16 @@ added: v0.1.21
-because the properties on the [`Error`][] object are non-enumerable:
+Only [enumerable "own" properties][] are considered. The `deepEqual()`
+implementation does not test the [`[[Prototype]]`][prototype-spec] of
+objects, attached symbols, or non-enumerable properties. (For those checks,
```
Just spotted this... can you actually remove the parens in this sentence. It can be reworded slightly as:

```
Only [enumerable "own" properties][] are considered. The `deepEqual()`
implementation does not test the [`[[Prototype]]`][prototype-spec] of objects,
attached symbols, or non-enumerable properties — for such checks, consider
using [assert.deepStrictEqual()][] instead.
```

Also, using `deepEqual()` in one sentence and `assert.deepStrictEqual()` in another is a bit inconsistent. Consider either using the `assert.` in both or omitting it from both.

joyeecheung replied:
Oh yes, I was thinking parens look a bit awkward too, the hyphen does the trick!
I'll use `assert.deepStrictEqual()` because it has a link already. Thanks!

joyeecheung commented:
@targos Thanks for the review, updated.

New CI: https://ci.nodejs.org/job/node-test-pull-request/6492/

The CITGM for the last code change has the same amount of failure as master (` (27 failures / ±0)`), not sure if we can rely on this though...

joyeecheung commented:
Hmm..don't know what else does this PR need to land? CITGM is `27 failures` and `±0` compared to its master.

addaleax commented:
@joyeecheung I don’t think there’s anything speaking against landing it. :)

joyeecheung commented:
@addaleax Alright, gonna land this tomorrow :) Thanks

Trott commented:
FYI, needs a rebase thanks to a PR I just landed.

Trott commented on lib/assert.js:
```diff
@@ -140,117 +140,176 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
 
-function objEquiv(a, b, strict, actualVisitedObjects) {
-  if (a === null || a === undefined || b === null || b === undefined)
+  // GH-7178. Ensure reflexivity of deepEqual with `arguments` objects.
```
This is probably a good time to change that GH reference to a full URL. Generally a good idea, but especially here since it's not in this repository, but in the archive one instead: 

https://github.com/nodejs/node-v0.x-archive/pull/7178

joyeecheung replied:
Yep, I'll update those links, thanks!

Trott commented on test/parallel/test-assert-deep.js:
```diff
@@ -0,0 +1,109 @@
+  new Int32Array([1]), // Int32Array
+  new Uint32Array([1]), // Uint32Array
+  Buffer.from([1]),
+  // Arguments {'0': '1'} is not here, see GH-7178
```
Same here, full URL would be helpful: https://github.com/nodejs/node-v0.x-archive/pull/7178

joyeecheung commented:
Rebased and squashed with full github links added.

New CI: https://ci.nodejs.org/job/node-test-pull-request/6583/
CITGM against master: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/599/
CITGM against this PR: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/600/

gibfahn commented:
@nodejs/citgm do these failures look reasonable?

joyeecheung commented:
This is weird, the CITGM against this PR is `(18 failures / -11)`, less failures than the master...is it just flakiness?

joyeecheung commented:
Not sure why CITGM got less errors with this PR, but according previous results I am pretty sure it's not related to the code change here. I am going to land this tomorrow if no more concerns arise...

jasnell commented:
CITGM and CI look fine. I've seen flakiness like that before.

joyeecheung commented:
Landed in https://github.com/nodejs/node/commit/562cf5a81c213f893a4787d1143e322c720e0eb9 and https://github.com/nodejs/node/commit/3d8379ae60ee2b6c57a55af7b62a878c920a793f, thanks!
